presto-orc/src/main/java/com/facebook/presto/orc/reader/MapFlatStreamReader.java

```diff
@@ -0,0 +1,429 @@
+            for (int valueIndex = 0; valueIndex < inMap.length; valueIndex++) {
+                if (inMap[valueIndex] == null || inMap[valueIndex][mapIndex]) {
+                    valueType.appendTo(valueBlocks[valueIndex], blockPositions[valueIndex], builder);
+                    blockPositions[valueIndex]++;
+                }
+            }
+        }
+
+        return builder.build();
+    }
+
+    private Block createKeyBlock(boolean[][] inMap, int totalEntryCount, int batchSize)
+    {
+        int[] ids = new int[totalEntryCount];
+        int idsIndex = 0;
+
+        for (int mapIndex = 0; mapIndex < batchSize; mapIndex++) {
+            for (int keyIndex = 0; keyIndex < inMap.length; keyIndex++) {
+                if (inMap[keyIndex] == null || inMap[keyIndex][mapIndex]) {
+                    ids[idsIndex++] = keyIndex;
+                }
+            }
+        }
+
+        return new DictionaryBlock(dictionaryBlock, ids);
+    }
+
+    private void openRowGroup()
+            throws IOException
+    {
+        presentStream = presentStreamSource.openStream();
+
+        for (InputStreamSource<BooleanInputStream> inMapStreamSource : inMapStreamSources) {
+            inMapStreams.add(inMapStreamSource.openStream());
+        }
+
+        rowGroupOpen = true;
+    }
+
+    @Override
+    public void startStripe(InputStreamSources dictionaryStreamSources, List<ColumnEncoding> encodings)
+            throws IOException
+    {
+        presentStreamSource = missingStreamSource(BooleanInputStream.class);
+
+        inMapStreamSources.clear();
+        valueStreamDescriptors.clear();
+        valueStreamReaders.clear();
+
+        ColumnEncoding encoding = encodings.get(baseValueStreamDescriptor.getStreamId());
+        checkState(
+                encoding.getAdditionalSequenceEncodings().isPresent(),
+                "The encoding for a flat map column does not contain additionalSequenceEncodings: " + encoding);
+        List<DwrfSequenceEncoding> additionalSequenceEncodings = encoding.getAdditionalSequenceEncodings().get();
+        // The ColumnEncoding with sequence ID 0 doesn't have any data associated with it
+        for (int sequence = 1; sequence <= additionalSequenceEncodings.size(); sequence++) {
+            inMapStreamSources.add(missingStreamSource(BooleanInputStream.class));
+
+            StreamDescriptor valueStreamDescriptor = copyStreamDescriptorWithSequence(baseValueStreamDescriptor, sequence);
+            valueStreamDescriptors.add(valueStreamDescriptor);
+
+            StreamReader valueStreamReader = StreamReaders.createStreamReader(valueStreamDescriptor, hiveStorageTimeZone, systemMemoryContext);
+            valueStreamReader.startStripe(dictionaryStreamSources, encodings);
+            valueStreamReaders.add(valueStreamReader);
+        }
+
+        dictionaryBlock = constructKeyBlock(additionalSequenceEncodings);
+        inMapVectors = new boolean[inMapStreamSources.size()][];
```

Comment:
Could these two fields change when moving to the next stripe?

Reply:
These fields should change every time a new stripe is started as the number of sequences (i.e. distinct keys) and the key values will change.